lib/internal/modules/esm/hooks.js

```diff
@@ -516,9 +460,187 @@ class Hooks {
+  /**
+   * The lock/unlock segment of the shared memory. Atomics require this to be a Int32Array. This
+   * segment is used to tell the main to sleep when the worker is processing.
+   * 0 -> main sleeps
+   * 1 -> main wakes up
+   */
+  #lock;
+  /**
+   * A MessagePort used to synchronously communicate with the worker.
+   */
+  #syncCommPort;
+
+  #isReady = false;
+
+  constructor() {
+    const { InternalWorker } = require('internal/worker');
+    MessageChannel ??= require('internal/worker/io').MessageChannel;
+
+    const lock = new SharedArrayBuffer(SHARED_MEMORY_BYTE_LENGTH);
+    this.#lock = new Int32Array(lock);
+    const syncCommChannel = new MessageChannel();
+    this.#syncCommPort = syncCommChannel.port1;
+
+    const worker = this.worker = new InternalWorker('internal/modules/esm/worker', {
+      stderr: false,
+      stdin: false,
+      stdout: false,
+      trackUnmanagedFds: false,
+      workerData: {
+        lock,
+        syncCommPort: syncCommChannel.port2,
+      },
+      transferList: [syncCommChannel.port2],
+    });
+    worker.unref(); // ! Allows the process to eventually exit.
+  }
+
+  #waitForWorker() {
+    if (!this.#isReady) {
+      const { kIsOnline } = require('internal/worker');
+      if (!this.worker[kIsOnline]) {
+        debug('wait for signal from worker');
+        AtomicsWait(this.#lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 0);
+        const response = receiveMessageOnPort(this.#syncCommPort);
+        const { preloadScripts } = this.#unwrapMessage(response);
+        this.#executePreloadScripts(preloadScripts);
+      }
+
+      this.#isReady = true;
+    }
+  }
+
+  #beforeExitHandler = () => {
+    const value = AtomicsOr(this.#lock, IDLENESS_OF_THREADS, MAIN_THREAD_IS_IDLE);
+    debug('beforeExit main thread', value, this.#lock);
+    if (value & WORKER_THREAD_IS_IDLE && AtomicsLoad(this.#lock, NUMBER_OF_INCOMING_MESSAGES) === 0) {
+      debug('send terminate order to worker');
+      // The worker has already reported waiting for us to die together.
+      this.worker.terminate();
+    } else {
+      // The worker is still in good health, and we just told it it's the
+      // only thing keeping us alive.
+      this.worker.ref();
+    }
+  };
+
+  async makeAsyncRequest(method, ...args) {
+    AtomicsAdd(this.#lock, NUMBER_OF_INCOMING_MESSAGES, 1);
+    this.#waitForWorker();
+
+    MessageChannel ??= require('internal/worker/io').MessageChannel;
+    const asyncCommChannel = new MessageChannel();
+
+    // Pass work to the worker.
+    debug('post async message to worker', { method, args });
+    this.#syncCommPort.postMessage({ method, args, port: asyncCommChannel.port2 }, [asyncCommChannel.port2]);
+    // In case the worker was bored, it might lock itself waiting for a signal.
+    AtomicsNotify(this.#lock, NUMBER_OF_INCOMING_MESSAGES);
+
+    if (AtomicsAdd(this.#lock, NUMBER_OF_INCOMING_ASYNC_RESPONSES, 1) === 0) {
+      process.on('beforeExit', this.#beforeExitHandler);
```

Comment:
This is fairly grokkable, but it's not quite clear (at least to me) why it's necessary to add the handler this way / at this point in time. Could whoever wrote this please add a code comment?